sdks/python/apache_beam/transforms/stats_test.py

```diff
@@ -345,5 +353,313 @@ def test_approximate_unique_globally_by_error_with_skewed_data(self):
+      assert_that(quantiles, equal_to([[0, 5, 6, 6, 6]]))
+
+  def test_zipfian_distribution(self):
+    with TestPipeline() as p:
+      data = []
+      for i in range(1, 1000):
+        data.append(int(1000 / i))
+      pc = p | Create(data)
+      quantiles = pc | beam.ApproximateQuantiles.Globally(5)
+      assert_that(quantiles, equal_to([[1, 1, 2, 4, 1000]]))
+
+  def test_alternate_comparator(self):
+    data = ["aa", "aaa", "aaaa", "b", "ccccc", "dddd", "zz"]
+    with TestPipeline() as p:
+      pc = p | Create(data)
+      quantiles = pc | beam.ApproximateQuantiles.Globally(3)
+      assert_that(quantiles, equal_to([["aa", "b", "zz"]]))
+
+    with TestPipeline() as p:
+      pc = p | Create(data)
+      comparator = lambda a, b: len(a) - len(b)  # order by length
+      quantiles = pc | beam.ApproximateQuantiles.Globally(3, comparator)
+      assert_that(quantiles, equal_to([["b", "aaa", "ccccc"]]))
+
+    with TestPipeline() as p:
+      pc = p | Create(data)
+      comparator = lambda a, b: len(a) - len(b)  # order by length
+      quantiles = pc | beam.ApproximateQuantiles.Globally(3, comparator,
+                                                          reverse=True)
+      assert_that(quantiles, equal_to([["ccccc", "aaa", "b"]]))
+
+    with TestPipeline() as p:
+      pc = p | Create(data)
+      quantiles = pc | beam.ApproximateQuantiles.Globally(3, key=len)
+      assert_that(quantiles, equal_to([["b", "aaa", "ccccc"]]))
+
+    with TestPipeline() as p:
+      pc = p | Create(data)
+      quantiles = pc | beam.ApproximateQuantiles.Globally(3, key=len,
+                                                          reverse=True)
+      assert_that(quantiles, equal_to([["ccccc", "aaa", "b"]]))
+
+  def test_global_display_data(self):
+    comparator = lambda a, b: a - b  # order by length
+    aq = beam.ApproximateQuantiles.Globally(3, comparator, key=len,
+                                            reverse=True)
+    data = DisplayData.create_from(aq)
+    expected_items = [
+        DisplayDataItemMatcher('num_quantiles', aq._num_quantiles),
+        DisplayDataItemMatcher('compare', aq._compare.__class__),
+        DisplayDataItemMatcher('key', aq._key.__class__),
+        DisplayDataItemMatcher('reverse', aq._reverse.__class__)
+    ]
+    hc.assert_that(data.items, hc.contains_inanyorder(*expected_items))
+
+  def test_perkey_display_data(self):
+    comparator = lambda a, b: a - b  # order by length
+    aq = beam.ApproximateQuantiles.PerKey(3, comparator, key=len, reverse=True)
+    data = DisplayData.create_from(aq)
+    expected_items = [
+        DisplayDataItemMatcher('num_quantiles', aq._num_quantiles),
+        DisplayDataItemMatcher('compare', aq._compare.__class__),
+        DisplayDataItemMatcher('key', aq._key.__class__),
+        DisplayDataItemMatcher('reverse', aq._reverse.__class__)
+    ]
+    hc.assert_that(data.items, hc.contains_inanyorder(*expected_items))
+
+
+def _build_quantilebuffer_test_data():
```

Comment:
Is there a bigger test that we can run to see if we produce right results in a larger data set. There is quite a bit of a math here and it would be good if we can test it.

Reply:
The test data is taken from **Munro-Paterson Algorithm** *(http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.6.6513&rep=rep1&type=pdf)* In which we create buffer size from 96 to 15,259 max.

By the way, it does not create any buffer, just calculates the sizes!

Reply:
Let's add a large one, that does not require a new test data file. How about something like 256-512?

Reply:
I didn't get your point which says, _"that does not require a new **test data file**"_  & ... _"**256-512**"_.

Can you please elaborate?

Reply:
Sorry for the confusion. I meant:
* Let's use a buffer size between 256-512. Pick any number that works for you in that range. How about 512 if that works?
* I wanted to keep the size small enough, so that it can be added right here in the stats_test.py file. Otherwise, we will probably need to have a separate data file to keep the test information. I wanted to avoid that case.

Reply:
Hi Ahmet.
As discussed, the `_build_quantilebuffer_test_data` method returns a list containing 25 items (which are the test scenarios):
```
[
[0.1, 100000, 11, 98], 
[0.1, 1000000, 14, 123], 
[0.1, 10000000, 17, 153], 
[0.1, 100000000, 21, 96], 
[0.1, 1000000000, 24, 120], 
[0.05, 100000, 11, 98], 
[0.05, 1000000, 14, 123], 
[0.05, 10000000, 17, 153], 
[0.05, 100000000, 20, 191], 
[0.05, 1000000000, 23, 239], 

[0.01, 100000, 9, 391], 

[0.01, 1000000, 11, 977], 
[0.01, 10000000, 14, 1221], 
[0.01, 100000000, 17, 1526], 
[0.01, 1000000000, 21, 954], 
[0.005, 100000, 8, 782], 
[0.005, 1000000, 11, 977], 
[0.005, 10000000, 14, 1221], 
[0.005, 100000000, 17, 1526], 
[0.005, 1000000000, 20, 1908], 
[0.001, 100000, 6, 3125], 
[0.001, 1000000, 9, 3907], 
[0.001, 10000000, 11, 9766], 
[0.001, 100000000, 14, 12208], 
[0.001, 1000000000, 17, 15259]
]
```

where each tuple holds epsilon, maxInputSize, numBuffers, bufferSize in that order. (These tuples are passed into test_efficiency and test_correctness as a parameterized.expand)

The range of values of bufferSize is 98 (smallest) to 15259 (largest).

The buffer size range you mentioned (256-512) is also represented in this generated test dataset.

And the `ApproximateQuantilesBufferTest` takes around 80 milliseconds to complete on my machine!